### ISSUE TITLE
Encryption

### ISSUE THREAD

**Issue by rain-1** [matrix-org/matrix-ircd#53]
Could it support encryption too?

Relevant links:

* https://matrix.org/docs/guides/e2e_implementation.html - I found the protocol spec 
* https://git.matrix.org/git/olm - Maybe binding to olm with ffi instead of implementing all the crypto in rust.
* https://github.com/wireapp/proteus
* https://github.com/ruma/ruma

**Comment by madduck** [comment]
It could. For that to happen, it would need to establish local storage to store keys etc., and someone needs to implement Megolm. Want to take a shot? I actually have some funds available to bootstrap such work…

**Comment by jcgruenhage** [comment]
Another relevant link: https://gitlab.gnome.org/World/metaolm
High level bindings to olm for rust, implementing megolm.

**Comment by madduck** [comment]
Now we just need someone to do it :)

**Comment by jcgruenhage** [comment]
I might have time later this year, not sure yet.

**Comment by termermc** [comment]
Curious if there has been anything regarding this

**Comment by jplatte** [comment]
No, there hasn't been much progress on matrix-ircd in general since last GSoC. If you are interested in working on it, please let me know and I will try to coordinate.

**Comment by codeandcreate** [comment]
For me there is much interest in it! :)

**Comment by jplatte** [comment]
So this codebase isn't in a great state. The async/await work from last GSoC has just been scrapped. It didn't work entirely and there were additional legal concerns that I was trying to resolve, but have just given up on.

I recommend anybody who is interested in the idea of matrix-ircd join [#matrix-rust-sdk:matrix.org](https://matrix.to/#/#matrix-rust-sdk:matrix.org) and talk about next steps there. There's already a minimal ircd for Matrix that builds on up-to-date libraries, but it's not actively being worked on. A continuation or fork of that project is likely the best way forward.

I might update the readme to mention about this in the future.

**Comment by termermc** [comment]
Where is the minimal IRCd that you mentioned?

**Comment by jplatte** [comment]
It's here: https://github.com/martinetd/matrirc/

**Comment by madduck** [comment]
@jplatte Thanks for the update!

I've been thinking about a Matrix IRCd for a long time, and as much as I wish to help out, I am not in the position to learn Rust currently. I also believe that the choice of Rust for matrix-ircd is one of the reasons that so few people ended up working on it.

Don't want to hijack this issue, but if you do put something into the README, please consider also adding a note that I am myself interested in starting this project from scratch using Python, and I'd love a couple of people to do that with. I am pretty proficient in Python and network protocols, and this would be my project to properly learn async…

**Comment by jplatte** [comment]
@madduck Can you provide a matrix room or just your user ID so I have something to link to?
